Add basic framework for carousel displays (aka `BeatmapCarouselV2`)

peppy opened the pull request:
PRing this early to get a sense of how people like the structure of `Carousel`, a new base class which attempts to separate display and model tracking concerns from `BeatmapCarousel`.

The implementation should be "feature complete", as in allowing a full implementation of the beatmap carousel (unless I've missed anything, which I probably have) with the assumption that a lot of the extra logic will be added external to the base `Carousel` class.

Rather than writing up a long PR description I'll opt to let the code speak for itself.

I've squashed most of my commits into logical steps I took as I gradually changed the direction of things to keep API surface as small as possible. You can check the full history [here](https://github.com/ppy/osu/compare/master...peppy:osu:song-select-refact?expand=1) if that's helpful.

- [x] Depends on https://github.com/ppy/osu/pull/31474
- [x] Depends on https://github.com/ppy/osu-framework/pull/6492

peppy committed:
Add basic selection support

peppy committed:
Add scroll position maintaining

peppy committed:
Stop exposing `CarouselItem` externally

peppy committed:
Generic type instead of raw `object`?

peppy committed:
Add basic implementation of new beatmap carousel

bdach commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -0,0 +1,444 @@
+        /// Setting this will ensure <see cref="CarouselItem.Selected"/> is set to <c>true</c> only on the matching <see cref="CarouselItem"/>.
+        /// Of note, if no matching item exists all items will be deselected while waiting for potential new item which matches.
+        /// </remarks>
+        public virtual object? CurrentSelection
```
I assume this is `object?` and not `T?` because the point is that you can select both the items inside this carousel and their headers. That however does give me pause, because the current carousel kinda maintains selection of *both* the beatmap set "header" and the beatmap "item".

We'll likely want to be able to tell which header is considered "selected" if its child item is selected, and vice versa. Not sure I have concrete proposals without experimenting, but it is a concern I want to raise nonetheless.

peppy replied:
Yeah. I Think `BeatmapCarousel` can track the two selections quite easily.

bdach commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -0,0 +1,444 @@
+                cancellationSource = cts;
+            }
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
```
Not sure how much of this you want to keep, but this and the bunch of log statements below look potentially removable.

peppy replied:
Kinda want to keep so we can track performance when deployed and users report potential risk issues.

At least initially.

bdach commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -0,0 +1,444 @@
+            }
+
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            IEnumerable<CarouselItem> items = new List<CarouselItem>(Items.Select(CreateCarouselItemForModel));
```
This sort of thing worries me. In memory profiling this is pretty huge, although it's probably not avoidable...

![1736764729](https://github.com/user-attachments/assets/492cc345-c5e3-4b6f-ad11-8c9862c33e3a)

However the more important and actionable part is that these are going to allocate on the update thread, are they not? This is before any `await`s, so this part of the method still executes synchronously. This should probably be on a background thread, at least after a certain size. Would also be nice to have cancellation support for this to possibly avoid trashing the heap too much.

peppy replied:
I guess we can move the transform (`Select`) part, but the copy needs to be locked or update thread. Just keeping simple for now.

bdach replied:
> but the copy needs to be locked or update thread

Well I guess this means this review thread just plain cannot be actioned upon because one way or another the update thread will have to bear the overhead of a copy of 200k+ objects in the worst cases...? In fact at this point you might as well not change anything because the `.Select()` here is lazy and thus is performing the functions of *both* a transform and a copy, so splitting it into two parts would only increase the overhead...

I dunno, I really think this *is* going to be a performance bottleneck we'll have to address at some point somehow.

At least maybe if the copy process could be aborted midway through it would be *some* solace...

peppy replied:
There's many ways we can improve of this, but it's already much faster than the old carousel since the majority is now async.

The copy overhead wasn't showing up in my profiling so I haven't addressed it yet.

bdach replied:
Well okay if you think it's not an issue I'm willing to go along with it for now.

bdach commented on osu.Game/Screens/SelectV2/ICarouselPanel.cs:
```diff
@@ -0,0 +1,23 @@
+        /// <summary>
+        /// The carousel item this drawable is representing. This is managed by <see cref="Carousel{T}"/> and should not be set manually.
+        /// </summary>
+        CarouselItem? Item { get; set; }
```
This setter being public-public looks rather bad, but there don't seem to be any elegant ways of hiding it (i.e. ones that don't involve casting), so maybe something I'd look away from...

peppy replied:
I tried multiple alternatives and couldn't find a nicer way without making things overly complicated.

bdach commented on osu.Game/Screens/SelectV2/BeatmapCarouselV2.cs:
```diff
@@ -0,0 +1,257 @@
+    {
+        public async Task<IEnumerable<CarouselItem>> Run(IEnumerable<CarouselItem> items, CancellationToken cancellationToken) => await Task.Run(() =>
+        {
+            // TODO: perform grouping based on FilterCriteria
```
Dunno if this was supposed to stay in here, but I will say that it would be nice to see an example of a group or sort change in the test scene to see if those interactions and operations work correctly.

peppy replied:
Was planning to leave this for a follow-up, but should be few LoC to add to this PR so I'll do that.

bdach reviewed (commented):
loose initial pass

peppy committed:
Merge branch 'master' into beatmap-carousel-v2

peppy committed:
Split out `BeatmapCarousel` classes and drop `V2` suffix

peppy committed:
Add difficulty, artist and title sort examples

Also:

- Adds hinting at grouping and header status of items
- Passes through criteria and prepare for grouping tests.
- Makes `Filters` list `protected` because naming clash with `Filter()`
  on `BeatmapCarousel`.

peppy committed:
Add animation and depth control

peppy committed:
Add loading overlay and refine filter flow

peppy commented:
I've added some sorting support, including the basic "difficulty split out panels" in line with what we have on `master`. Also more animation and loading display.

Won't add any more features on this branch unless requested.

peppy committed:
Fix inspections that don't show in rider

bdach commented on osu.Game/Screens/SelectV2/Carousel.cs:
```diff
@@ -0,0 +1,454 @@
+
+            updateSelection();
+
+            void log(string text) => Logger.Log($"Carousel[op {cts.GetHashCode().ToString().Substring(0, 5)}] {stopwatch.ElapsedMilliseconds} ms: {text}");
```
As per the other test failure, this fails if `GetHashCode()` returns a number less than 5 digits long. I'd honestly just remove the substring because everything else will be more annoying to do?

bdach reviewed (commented):
Can you check on the test failures in `TestScrollPositionVelocityMaintained`? I'm not sure what to make of it in regards to expectations...

peppy committed:
Fix substring of `GetHashCode` potentially failing

peppy commented:
> Can you check on the test failures in TestScrollPositionVelocityMaintained? I'm not sure what to make of it in regards to expectations...

I think this broke when I added animations.. will see what I can do. Maybe just checking within 1 pixel will be enough to fix it.

peppy committed:
Move animation handling to `Carousel` implementation to better handle add/removes

With the animation logic being external, it was going to make it very
hard to apply the scroll offset when a new panel is added or removed
before the current selection.

There's no real reason for the animations to be local to beatmap
carousel. If there's a usage in the future where the animation is to
change, we can add more customisation to `Carousel` itself.

peppy committed:
Update framework

bdach commented on osu.Game/Screens/SelectV2/BeatmapCarouselFilterSorting.cs:
```diff
@@ -0,0 +1,65 @@
+        {
+            var criteria = getCriteria();
+
+            return items.OrderDescending(Comparer<CarouselItem>.Create((a, b) =>
```
I was a bit surprised by this being `.OrderDescending()` rather than just `Order()` but I suppose it doesn't matter too much for the time being.

bdach reviewed (approved):
Let's get this in and keep the train rolling :+1:
